Fix #4167: Update answer summarizers job to properly handle answers corresponding to deleted explorations

BenHenning opened the pull request:
Update InteractionAnswerSummariesMRJobManager to properly handle answers corresponding to deleted explorations corresponding to deleted explorations in the VERSION_ALL aggregation case.

**Checklist**
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes.
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR is assigned to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue.

BenHenning committed:
Update InteractionAnswerSummariesMRJobManager to properly handle answers
corresponding to deleted explorations in the VERSION_ALL aggregation
case.

BenHenning commented:
Currently working to verify this in a prod instance to ensure it's fixed.

seanlip commented on core/domain/stats_jobs_continuous.py:
```diff
@@ -611,8 +611,9 @@ def reduce(key, stringified_values):
-            if state_name in exp.states:
+            exp = exp_services.get_exploration_by_id(
+                exploration_id, strict=False)
+            if exp and state_name in exp.states:
```
Consider just returning if we discover that exp is None -- seems more resilient to future change, since we can then assume from this point on that the exploration exists and is fully valid.

Also, maybe move this check-and-return out of the "if" clause? Even if previous exp versions are available I don't see the point in computing any models for particular versions if the exploration has been deleted.

BenHenning replied:
Done.

BenHenning commented:
Verified this fixes the issue in prod.

BenHenning committed:
Address reviewer comments (ensure the exploration exists for answers
regardless of whether answers for all versions of an exploration or a
particular version are being aggregated).

BenHenning commented:
Thanks for the review!

seanlip commented on core/domain/stats_jobs_continuous.py:
```diff
@@ -623,6 +628,8 @@ def reduce(key, stringified_values):
                         'since the latest exploration version is %s' % (
                             latest_version, exp.version))
                     versions = []
+            elif not exp:
```
This is no longer necessary due to the check in line 612, right?

BenHenning replied:
Ah yes, correct. Removed.

seanlip commented on core/domain/stats_jobs_continuous.py:
```diff
@@ -606,12 +606,17 @@ def reduce(key, stringified_values):
+        # Ensure the exploration corresponding to these answers exists.
+        exp = exp_services.get_exploration_by_id(
+            exploration_id, strict=False)
+        if not exp:
```
Should be: `if exp is None`

BenHenning replied:
Done.

BenHenning committed:
Address second round of review comments.
